File: ecdsa-verify-dart/ecdsa_verify.dart
export 'ecdsa_verify.dart';

import 'dart:math';

BigInt? modularInverse(BigInt a, BigInt n) {
    if (n == BigInt.zero || n == BigInt.one) {
        return null;
    }
    BigInt t = BigInt.zero;
    BigInt newt = BigInt.one;
    BigInt r = n;
    BigInt newr = a;
    while (newr != BigInt.zero) {
        BigInt quotient = r ~/ newr;
        var tmp = t;
        t = newt;
        newt = tmp - quotient * newt;
        tmp = r;
        r = newr;
        newr = tmp - quotient * newr;
    }
    if (r > BigInt.one) {
        return null;
    }
    if (t < BigInt.zero) {
        t += n;
    }
    return t;
}

File: ecdsa-verify-ruby/ecdsa_verify.rb
def modular_inverse(a, n)
    return nil if n == 0 || n == 1

    t, newt = 0, 1
    r, newr = n.clone, a.clone
    while newr != 0
        quotient = r / newr
        t, newt = newt, t - quotient * newt
        r, newr = newr, r - quotient * newr
    end

    return nil if r > 1
    t += n if t < 0

    t
end
